integration-tests/src/tests/client/features/nonrefundable_transfer.rs

```diff
@@ -0,0 +1,346 @@
+};
+use near_primitives_core::account::{AccessKey, AccessKeyPermission};
+use nearcore::config::GenesisExt;
+use nearcore::test_utils::TestEnvNightshadeSetupExt;
+use nearcore::NEAR_BASE;
+
+/// Default sender to use in tests of this module.
+fn sender() -> AccountId {
+    "test0".parse().unwrap()
+}
+
+/// Default receiver to use in tests of this module.
+fn receiver() -> AccountId {
+    "test1".parse().unwrap()
+}
+
+/// Default signer (corresponding to the default sender) to use in tests of this module.
+fn signer() -> InMemorySigner {
+    InMemorySigner::from_seed(sender(), KeyType::ED25519, "test0")
+}
+
+/// Creates a test environment using given protocol version (if some).
+fn setup_env_with_protocol_version(protocol_version: Option<ProtocolVersion>) -> TestEnv {
+    let mut genesis = Genesis::test(vec![sender(), receiver()], 1);
+    if let Some(protocol_version) = protocol_version {
+        genesis.config.protocol_version = protocol_version;
+    }
+    TestEnv::builder(ChainGenesis::new(&genesis))
+        .real_epoch_managers(&genesis.config)
+        .nightshade_runtimes(&genesis)
+        .build()
+}
+
+/// Creates a test environment using default protocol version.
+fn setup_env() -> TestEnv {
+    setup_env_with_protocol_version(None)
+}
+
+fn get_nonce(env: &mut TestEnv, signer: &InMemorySigner) -> u64 {
+    let request = QueryRequest::ViewAccessKey {
+        account_id: signer.account_id.clone(),
+        public_key: signer.public_key.clone(),
+    };
+    match env.query_view(request).unwrap().kind {
+        QueryResponseKind::AccessKey(view) => view.nonce,
+        _ => panic!("wrong query response"),
+    }
+}
+
+fn account_exists(env: &mut TestEnv, account_id: AccountId) -> bool {
+    let request = QueryRequest::ViewAccount { account_id };
+    env.query_view(request).is_ok()
+}
+
+fn execute_transaction_from_actions(
+    env: &mut TestEnv,
+    actions: Vec<Action>,
+    signer: &InMemorySigner,
+    receiver: AccountId,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    let tip = env.clients[0].chain.head().unwrap();
+    let nonce = get_nonce(env, signer);
+    let tx = SignedTransaction::from_actions(
+        nonce + 1,
+        signer.account_id.clone(),
+        receiver,
+        signer,
+        actions,
+        tip.last_block_hash,
+    );
+    let tx_result = env.execute_tx(tx);
+    let height = env.clients[0].chain.head().unwrap().height;
+    for i in 0..2 {
+        env.produce_block(0, height + 1 + i);
+    }
+    tx_result
+}
+
+/// Submits a transfer (either regular or non-refundable).
+///
+/// This methods checks that the balance is subtracted from the sender and added
+/// to the receiver, if the status was ok. No checks are done on an error.
+fn exec_transfer(
+    env: &mut TestEnv,
+    signer: InMemorySigner,
+    receiver: AccountId,
+    deposit: Balance,
+    nonrefundable: bool,
+    account_creation: bool,
+    implicit_account_creation: bool,
+    deploy_contract: bool,
```

Comment:
The number of boolean flags here all in a row make the tests a little hard to read in my opinion. Maybe it would be nicer (though more verbose) to pass a test config struct of some kind instead. Then at the callsite there would be something like

```rust
exec_transfer(env, signer, receiver, deposit, TransferConfig { nonrefundable: true, account_creation: false, ... })
```